torch/csrc/jit/passes/onnx/deduplicate_initializers.cpp

```diff
@@ -0,0 +1,92 @@
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/onnx/deduplicate_initializers.h>
+#include <torch/csrc/jit/passes/onnx/helper.h>
+
+#include <c10/util/irange.h>
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+void DeduplicateInitializers(
+    std::shared_ptr<Graph>& g,
+    ValueToParamPairMap& valsToParamsMap,
+    int opset_version,
+    bool (*comp)(at::Tensor&, at::Tensor&)) {
+  auto is_same_tensor_as = [&valsToParamsMap, comp](Value* v1) {
+    return [&valsToParamsMap, v1, comp](Value* v2) {
+      if ((valsToParamsMap.find(v1) == valsToParamsMap.end()) ||
+          (valsToParamsMap.find(v2) == valsToParamsMap.end())) {
+        return false;
+      }
+      auto iv1 = valsToParamsMap.find(v1)->second.second;
+      auto iv2 = valsToParamsMap.find(v2)->second.second;
+      if (!iv1.isTensor() || !iv2.isTensor()) {
+        return false;
+      }
+      auto t1 = iv1.toTensor();
+      auto t2 = iv2.toTensor();
+      return comp(t1, t2);
+    };
+  };
+  std::vector<Value*> uniqueVals;
+  std::vector<size_t> inputsIndicesToRemove;
+  auto b = g->block();
+
+  for (auto i : c10::irange(b->inputs().size())) {
+    auto v = g->inputs().at(i);
+    if (valsToParamsMap.find(v) == valsToParamsMap.end()) {
+      // Skip model inputs
+      continue;
+    }
+    auto it = std::find_if(
+        uniqueVals.begin(), uniqueVals.end(), is_same_tensor_as(v));
+    if (it == uniqueVals.end()) {
+      uniqueVals.emplace_back(v);
+    } else {
+      inputsIndicesToRemove.emplace_back(i);
+      v->replaceAllUsesWith(*it);
+    }
+  }
+  for (auto it = inputsIndicesToRemove.rbegin();
+       it != inputsIndicesToRemove.rend();
+       ++it) {
+    valsToParamsMap.erase(g->inputs().at(*it));
+    g->eraseInput(*it);
+  }
+}
+
+bool DeduplicateInitializersByDataPtr(at::Tensor& t1, at::Tensor& t2) {
+  return t1.sizes().equals(t2.sizes()) && t1.strides().equals(t2.strides()) &&
+      (t1.data_ptr() == t2.data_ptr());
+}
+
+bool DeduplicateInitializersByValue(at::Tensor& t1, at::Tensor& t2) {
+  return t1.dtype() == t2.dtype() && t1.equal(t2);
+}
+
+void DeduplicateInitializers(
+    std::shared_ptr<Graph>& g,
+    std::map<std::string, IValue>& paramsDict,
+    bool is_train,
+    int opset_version) {
+  auto valsToParamsMap = buildValueToParamsMap(g->block(), paramsDict);
+  // ONNX spec does not support parameters with shared memory.
+  // This pass de-duplicate those parameters. Training is not affected.
+  DeduplicateInitializers(
+      g, valsToParamsMap, opset_version, DeduplicateInitializersByDataPtr);
+  if (!is_train) {
+    // More aggressive parameters de-duplication based on tensor values.
```

Comment:
Can you please rewrite the comment to explain why `ByValue` is not safe for training?
Something like:
```When training, a runtime may <fill me in>```